test/integration/service/loadbalancer_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"context"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/test/integration/framework"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// Test_ServiceLoadBalancerAllocateNodePorts tests that a Service with spec.allocateLoadBalancerNodePorts=false
+// does not allocate node ports for the Service.
+func Test_ServiceLoadBalancerDisableAllocateNodePorts(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceLBNodePortControl, true)()
+
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	_, server, closeFn := framework.RunAMaster(masterConfig)
+	defer closeFn()
+
+	config := restclient.Config{Host: server.URL}
+	client, err := clientset.NewForConfig(&config)
+	if err != nil {
+		t.Fatalf("Error creating clientset: %v", err)
+	}
+
+	ns := framework.CreateTestingNamespace("test-service-allocate-node-ports", server, t)
+	defer framework.DeleteTestingNamespace(ns, server, t)
+
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-123",
+		},
+		Spec: corev1.ServiceSpec{
+			Type:                          corev1.ServiceTypeLoadBalancer,
+			AllocateLoadBalancerNodePorts: utilpointer.BoolPtr(false),
+			Ports: []corev1.ServicePort{{
+				Port: int32(80),
+			}},
+			Selector: map[string]string{
+				"foo": "bar",
+			},
+		},
+	}
+
+	service, err = client.CoreV1().Services(ns.Name).Create(context.TODO(), service, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating test service: %v", err)
+	}
+
+	foundNodePorts := false
+	for _, port := range service.Spec.Ports {
+		if port.NodePort > 0 {
+			foundNodePorts = true
+		}
+	}
+
+	if foundNodePorts {
+		t.Error("found node ports when none was expected")
+	}
+}
+
+// Test_ServiceLoadBalancerSwitchToDeallocatedNodePorts test that switching a Service
+// to spec.allocateLoadBalancerNodePorts=false, does not de-allocate existing node ports.
+func Test_ServiceLoadBalancerEnableThenDisableAllocatedNodePorts(t *testing.T) {
```

Comment:
What happens if I have a NodePort service and I mutate to LoadBalancer without NodePort enables?
Same with LoadBalancer type without node ports that mutates to NodePort

Reply:
> What happens if I have a NodePort service and I mutate to LoadBalancer without NodePort enables?

This isn't possible today since Type=NodePorts is always allocated node ports. The new field `service.spec.allocateLoadBalancerNodePorts` only applies for Type=LoadBalancer. 

You can however, go from NodePort -> LoadBalancer and set `service.spec.allocateLoadBalancerNodePorts=false`. In this case nothing happens because that field does not de-allocate existing node ports, it only prevents new ones from being allocated.

Reply:
In general though, I feel like we should have an integration test suite that just tests various behaviors when switching between Service types. I would be willing to work on that as a follow-up to this PR.

Reply:
thanks for the clarifications
> it only prevents new ones from being allocated.